src/meta/meta_ckpt.c

```diff
@@ -1347,18 +1515,255 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
       txn->snap_min, txn->snap_max, txn->snapshot_count,
       __wt_timestamp_to_string(txn_global->oldest_timestamp, ts_string[0]),
       __wt_timestamp_to_string(txn_global->meta_ckpt_timestamp, ts_string[1]),
-      S2C(session)->base_write_gen);
+      conn->base_write_gen);
+
+    /* Record the base write gen in metadata as part of full checkpoints. */
 
-    /* Record the base write gen in metadata as part of checkpoint */
-    WT_ERR(__wt_buf_fmt(
-      session, buf, WT_SYSTEM_BASE_WRITE_GEN "=%" PRIu64, S2C(session)->base_write_gen));
-    WT_ERR(__wt_metadata_update(session, WT_SYSTEM_BASE_WRITE_GEN_URI, buf->data));
+    if (full) {
```

Comment:
Maybe a basic doubt on named checkpoints? Named checkpoints are not full checkpoints?

Reply:
This "full" is passed from __txn_checkpoint where it specifically means "the user didn't provide a list of trees to checkpoint". It is allowed for the user to give a name and a list of trees that skips some files. Right now because of the history store that's a doomed thing to do and there's a strong caution against it in the docs, but in the world where each tree has its own history it becomes sensible again. I've added a note here since it's neither obvious nor local to this file. (note is in 48a4977)

Reply:
@kommiharibabu, I'm not 100% sure that is what you were asking -- the goal here is for "named" and "unnamed" checkpoints to be identical except "named" checkpoints aren't automatically removed. This includes a change that MDB opening the "last" checkpoint in the tree might get either a named or unnamed checkpoint, and it shouldn't be able to tell the difference.

Reply:
@keithbostic My confusion starts with the following code in txn_ckpt.c
`    if (full || name != NULL)`
          `WT_ERR(__wt_meta_sysinfo_set(session, full, name, namelen));`
I got that doubt that it is possible that we don't have a checkpoint on-disk that is not a full checkpoint?

Reply:
Thanks, @keithbostic and @sauclovian-wt.

Reply:
Sorry for not being clear on my confusion. My afraid was that we are not saving the `WT_SYSTEM_BASE_WRITE_GEN` in the metadata for the named checkpoints, I thought that whether it can cause any problems during the recovery if we used that named checkpoint as a recovery checkpoint?

Further checking I found that it may not cause any problems as the `conn->base_write_gen` doesn't change and at least one latest value is available in the metadata as part of the recovery checkpoint. 

Why can't we save the `WT_SYSTEM_BASE_WRITE_GEN` even if it is not a full checkpoint? Saving shouldn't cause any problems.

Reply:
Ah! Ok. The way it works now is that it saves the "most recent checkpoint info" metadata if `full` is set, and the "named checkpoint info metadata for this name" if there's a name, and if both are true it saves both. 

The way it worked before this changeset is that it saved "the most recent checkpoint info" metadata if `full` is set. It didn't call `__wt_meta_sysinfo_set` at all otherwise, so the diffs are slightly misleading. Unless I've screwed up badly somewhere, all of "the most recent checkpoint info" metadata is written under the same circumstances it was before.

We recover to the most recent full checkpoint, whether or not it's named; that hasn't changed. This means that for named full checkpoints we write the metadata twice, once as the most recent checkpoint and once under the checkpoint's name. This is why I changed the behavior of opening "WiredTigerCheckpoint" to always read the most recent checkpoint (named or unnamed) instead of only the most recent unnamed checkpoint: otherwise the snapshot info isn't available.

The `WT_SYSTEM_BASE_WRITE_GEN` value is associated with the most recent checkpoint and used for recovery; I'd expect that if we updated it during a non-full checkpoint (named or otherwise) bad things would happen.

Reply:
The `WT_SYSTEM_BASE_WRITE_GEN` is updated from the connection `base_write_gen` and it gets modified only during the recovery phase. So it shouldn't be a problem whether we update it or not with every checkpoint because as part of the recovery we already updated it to the latest value with the recovery checkpoint.